### PR TITLE
fix proper keywords.

### PR DESCRIPTION
**Pull request by aakashrajput**
added proper meta keywords

### CODE REVIEW

**Comment by aakashrajput** [comment]
@googlebot I signed it!

**Review by shivaylamba — APPROVED** [review]

File: _layouts/base.html [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ---
 layout: default
 ---
+{% assign keywords = "google,asylo,opensource framework,flexible framework,opensource application,asylo-site" %}
```
Comment (KeithMoyer) [inline _layouts/base.html]: I admit I'm not a Jekyll expert. Does this not override values set in the page front matter? I expected default values (that can be overridden by front matter) to be declared something similar to https://jekyllrb.com/docs/configuration/front-matter-defaults/

File: index.html [hunk 2]
```diff
@@ -1,6 +1,7 @@
 ---
 layout: landing
 title: Asylo
+keywords: google,asylo,opensource framework,flexible framework,opensource application,asylo-site
```
Comment (KeithMoyer) [inline index.html]: Why have explicit keywords here when it should inherit like the other pages?

File: _layouts/base.html (hunk 1 above)
Comment (KeithMoyer) [inline _layouts/base.html]: fitting "enclave" in the list would be good.

**Review by KeithMoyer — CHANGES_REQUESTED** [review]
Could you give some context on what prompted these changes? I like the idea of making this front-matter-settable, but I'm curious what led you to make this change.

File: index.html (hunk 2 above)
Comment (aakashrajput) [reply index.html]: for front matter defaults we need to add it on the index.html page

File: _layouts/base.html (hunk 1 above)
Comment (aakashrajput) [reply _layouts/base.html]: done

Comment (aakashrajput) [reply _layouts/base.html]: need to remove from this and add it to the front matter defaults (ie to the index.html)

Comment (KeithMoyer) [reply _layouts/base.html]: Do you have a reference/documentation for that being the case? It doesn't seem to work. With your current PR contents, all pages except for index.html have empty keywords.

**Comment by KeithMoyer** [comment]
Still curious what brought this to your attention.

**Comment by aakashrajput** [comment]
> Still curious what brought this to your attention.

The old keywords were not proper so I thought of changing them so used this method to update keywords so that it can easily be available on other pages for the better search option.

**Comment by shivaylamba** [comment]
@aakashrajput I think the old keywords were adequate and now since keywords are anyways deprecated and not used for SEO, so I don't think there is need for addition of further keywords. But I guess it's ok to add few other keywords. But they don't provide any further use case as such.

**Comment by KeithMoyer** [comment]
> The old keywords were not proper so I thought of changing them so used this method to update keywords so that it can easily be available on other pages for the better search option.

Thanks for the clarification. I agree the existing keywords aren't really of canonical form. I was just wondering if you ran across this via some tool analyzing the site or just inspecting the code out of interest (or found it some other way). This is just to satisfy my own curiosity.

The bigger issue is that these changes don't seem to work properly. There are instructions in the project README.md on how to run a copy of the site locally, to help get all the machinery and syntax right.

**Comment by aakashrajput** [comment]
> > The old keywords were not proper so I thought of changing them so used this method to update keywords so that it can easily be available on other pages for the better search option.
> 
> Thanks for the clarification. I agree the existing keywords aren't really of canonical form. I was just wondering if you ran across this via some tool analyzing the site or just inspecting the code out of interest (or found it some other way). This is just to satisfy my own curiosity.
> 
> The bigger issue is that these changes don't seem to work properly. There are instructions in the project README.md on how to run a copy of the site locally, to help get all the machinery and syntax right.

I was inspecting the code but also I ran the site locally using Docker.

Screen Shot 2019-10-17 at 1 16 32 AM

attached my source of locally working example with the keywords. 
I agree the old one I pushed was not proper so I modified them again.
https://github.com/google/asylo-site/pull/12/commits/6b31b3d1f857e53857a1cffb1ab89d8b892a985c

**Comment by KeithMoyer** [comment]
> > > The old keywords were not proper so I thought of changing them so used this method to update keywords so that it can easily be available on other pages for the better search option.
> > 
> > 
> > Thanks for the clarification. I agree the existing keywords aren't really of canonical form. I was just wondering if you ran across this via some tool analyzing the site or just inspecting the code out of interest (or found it some other way). This is just to satisfy my own curiosity.
> > The bigger issue is that these changes don't seem to work properly. There are instructions in the project README.md on how to run a copy of the site locally, to help get all the machinery and syntax right.
> 
> I was inspecting the code but also I ran the site locally using Docker.
> 
> Screen Shot 2019-10-17 at 1 16 32 AM
> 
> attached my source of locally working example with the keywords.
> I agree the old one I pushed was not proper so I modified them again.
> [6b31b3d](https://github.com/google/asylo-site/commit/6b31b3d1f857e53857a1cffb1ab89d8b892a985c)

6b31b3d is what I tested to see that it does not work properly. The new keywords show up on the home page, but your changes make them disappear entirely on all other pages.

```html
<meta name="title" content="Asylo API Overview">
<meta name="description" content="A general overview of Asylo application development">
<meta name="keywords" content="">
```

**Comment by aakashrajput** [comment]
> > > > The old keywords were not proper so I thought of changing them so used this method to update keywords so that it can easily be available on other pages for the better search option.
> > > 
> > > 
> > > Thanks for the clarification. I agree the existing keywords aren't really of canonical form. I was just wondering if you ran across this via some tool analyzing the site or just inspecting the code out of interest (or found it some other way). This is just to satisfy my own curiosity.
> > > The bigger issue is that these changes don't seem to work properly. There are instructions in the project README.md on how to run a copy of the site locally, to help get all the machinery and syntax right.
> > 
> > 
> > I was inspecting the code but also I ran the site locally using Docker.
> > Screen Shot 2019-10-17 at 1 16 32 AM
> > attached my source of locally working example with the keywords.
> > I agree the old one I pushed was not proper so I modified them again.
> > [6b31b3d](https://github.com/google/asylo-site/commit/6b31b3d1f857e53857a1cffb1ab89d8b892a985c)
> 
> [6b31b3d](https://github.com/google/asylo-site/commit/6b31b3d1f857e53857a1cffb1ab89d8b892a985c) is what I tested to see that it does not work properly. The new keywords show up on the home page, but your changes make them disappear entirely on all other pages.
> 
> ```
> <meta name="title" content="Asylo API Overview">
> <meta name="description" content="A general overview of Asylo application development">
> <meta name="keywords" content="">
> ```

First, I was using keywords as `pages.keywords` to display it just on the home page but as we don't have different meta keywords for different pages. I assigned keywords so it can be used by all the other pages, check it once.

**Comment by shivaylamba** [comment]
> > > > > The old keywords were not proper so I thought of changing them so used this method to update keywords so that it can easily be available on other pages for the better search option.
> > > > 
> > > > 
> > > > Thanks for the clarification. I agree the existing keywords aren't really of canonical form. I was just wondering if you ran across this via some tool analyzing the site or just inspecting the code out of interest (or found it some other way). This is just to satisfy my own curiosity.
> > > > The bigger issue is that these changes don't seem to work properly. There are instructions in the project README.md on how to run a copy of the site locally, to help get all the machinery and syntax right.
> > > 
> > > 
> > > I was inspecting the code but also I ran the site locally using Docker.
> > > Screen Shot 2019-10-17 at 1 16 32 AM
> > > attached my source of locally working example with the keywords.
> > > I agree the old one I pushed was not proper so I modified them again.
> > > [6b31b3d](https://github.com/google/asylo-site/commit/6b31b3d1f857e53857a1cffb1ab89d8b892a985c)
> > 
> > 
> > [6b31b3d](https://github.com/google/asylo-site/commit/6b31b3d1f857e53857a1cffb1ab89d8b892a985c) is what I tested to see that it does not work properly. The new keywords show up on the home page, but your changes make them disappear entirely on all other pages.
> > ```
> > <meta name="title" content="Asylo API Overview">
> > <meta name="description" content="A general overview of Asylo application development">
> > <meta name="keywords" content="">
> > ```
> 
> First, I was using keywords as `pages.keywords` to display it just on the home page but as we don't have different meta keywords for different pages. I assigned keywords so it can be used by all the other pages, check it once.

@aakashrajput I tested the latest commit 6b31b3d It is working fine now, for all pages.

**Comment by KeithMoyer** [comment]
> First, I was using keywords as `pages.keywords` to display it just on the home page but as we don't have different meta keywords for different pages. I assigned keywords so it can be used by all the other pages, check it once.

The latest commit (88ebe4d) does not appear to work for any pages.  They all have:
```html
<meta name="keywords" content="">
```

In your replies and your commit messages, you've repeated that putting keywords on index.html makes them accessible on all pages. However, the documentation I linked and the testing I've done suggest that is not true. As I asked before, where are you seeing that that should be the case?

> @aakashrajput I tested the latest commit 6b31b3d It is working fine now, for all pages.

In what way are you testing? 6b31b3d is close, in that it allows pages to specify their own keywords. But, there is no default and all pages that don't explicitly specify keywords get none.

**Comment by shivaylamba** [comment]
@KeithMoyer With the current commit, in which the keywords are only mentioned in _layouts/base.html, the keywords are visible on all pages. So why is there a need to include them in index.html, when they are visible in the rest of the pages as well, if we are keeping a unified set of keywords for all pages?

**Comment by KeithMoyer** [comment]
> @KeithMoyer With the current commit, in which the keywords are only mentioned in _layouts/base.html, the keywords are visible on all pages. So why is there a need to include them in index.html, when they are visible in the rest of the pages as well, if we are keeping a unified set of keywords for all pages?

With the current commit, the keywords do not show up on any of the rendered pages (not even on the home page).

Before this PR, the keywords were listed only in `_layouts/base.html`, which I am fine with because we currently use the same keywords on all pages.

This PR appears to be attempting to make them customizable per-page. For me, this isn't a requirement, but if @aakashrajput is putting in the work, there is at least no harm in it. The added flexibility _could_ be nice in the future, in case we want to use different tags on some pages.

Whether or not each page can customize the keywords, keywords should still show up on all of the pages. That has been the issue with the commits so far. In my testing, the "the keywords are visible on all pages" statement does not hold true.

**Comment by aakashrajput** [comment]
> > @KeithMoyer With the current commit, in which the keywords are only mentioned in _layouts/base.html, the keywords are visible on all pages. So why is there a need to include them in index.html, when they are visible in the rest of the pages as well, if we are keeping a unified set of keywords for all pages?
> 
> With the current commit, the keywords do not show up on any of the rendered pages (not even on the home page).
> 
> Before this PR, the keywords were listed only in `_layouts/base.html`, which I am fine with because we currently use the same keywords on all pages.
> 
> This PR appears to be attempting to make them customizable per-page. For me, this isn't a requirement, but if @aakashrajput is putting in the work, there is at least no harm in it. The added flexibility _could_ be nice in the future, in case we want to use different tags on some pages.
> 
> Whether or not each page can customize the keywords, keywords should still show up on all of the pages. That has been the issue with the commits so far. In my testing, the "the keywords are visible on all pages" statement does not hold true.

is my latest commit also not showing keywords on all the pages?

**Comment by KeithMoyer** [comment]
> is my latest commit also not showing keywords on all the pages?

That's correct. Commit 88ebe4d appears to show keywords on no pages.

**Comment by aakashrajput** [comment]
> > is my latest commit also not showing keywords on all the pages?
> 
> That's correct. Commit [88ebe4d](https://github.com/google/asylo-site/commit/88ebe4d4662728f8889c52320d499cd4fd51964b) appears to show keywords on no pages.

In the last commit, I assigned the keywords at the wrong place fixed now. Next we can work on creating different keywords for different pages.

File: index.html [hunk 3]
```diff
@@ -103,4 +103,3 @@ <h2>Ready to get started?</h2>
         padding-top: 2.8rem;
     }
 </style>
-
```
Comment (KeithMoyer) [inline index.html]: The only change to this file is removing an empty line. Please revert this change.

File: _layouts/base.html [hunk 4]
```diff
@@ -2,6 +2,7 @@
 layout: default
 ---
 {% include home.html %}
+{% assign keywords = "google,asylo,opensource framework,flexible framework,opensource application,asylo-site,enclave,enclave application" %}
```
Comment (KeithMoyer) [inline _layouts/base.html]: With this approach, pages cannot override keywords (which seemed to be the intent in the original changes). All pages get exactly these keywords. This is fine, but if this PR is limited to updating the existing global list of keywords, I think I'd prefer just updating the list of keywords where they were instead of creating a variable (that doesn't have a way for it to change) for it.

**Review by KeithMoyer — CHANGES_REQUESTED** [review]

File: _layouts/base.html (hunk 4 above)
Comment (aakashrajput) [reply _layouts/base.html]: with this commit we can easily update the global lists of keywords, the thing is if we want to use different keywords for different pages like adding a new keyword 'about' for about page, we can do that within the about page and can have a statement in base.html that can assign what keywords we need to choose for the respected pages.

But for now we can have these global lists of keywords that will show on all the pages and if we want some new keywords we can easily add them in the list.

Comment (KeithMoyer) [reply _layouts/base.html]: > with this commit we can easily update the global lists of keywords

Yes, but that was also true before this change.

> if we want to use different keywords for different pages like adding a new keyword 'about' for about page, we can do that within the about page and can have a statement in base.html that can assign what keywords we need to choose for the respected pages.

This PR doesn't do anything to make it so that the about page can set its own keywords. Like I said before, that's fine, though.

If what this PR accomplishes is updating the hard-coded list of keywords to a new hard-coded list, then I'd prefer just updating the list where it currently is. Moving them up to a variable doesn't match how the other hard-coded values are handled on this page and doesn't allow for anything new in itself.

If the argument is that there should be no meaningful hard-coded values, then that could be a separate conversation and evenly applied throughout.  If the argument is that it will need to be templated in order to plug in content from page front-matter, let's address that if/when we add the ability for the front-matter to control the keywords.

**Review by KeithMoyer — CHANGES_REQUESTED** [review]

**Comment by aakashrajput** [comment]
Now in my latest commit https://github.com/google/asylo-site/pull/12/commits/969e33017cb7c12d154db1df3a2074ae76699c4d 

I added different keywords for different pages, now if the respected page has any keywords it will display that else it will display the default keywords which are in the global lists.

**Review by shivaylamba — APPROVED** [review]
This commit is working fine. This method like the one controlling description for each page, lets each page their custom keywords as well.

File: index.html (hunk 3 above)
Comment (KeithMoyer) [inline index.html]: Only change to this file is removing a line.

File: _layouts/base.html [hunk 5]
```diff
@@ -3,6 +3,13 @@
 ---
 {% include home.html %}
 
+{% if page.keywords == nil %}
```
Comment (KeithMoyer) [inline _layouts/base.html]: Please move this logic down with the similar logic for description

File: _about/about.md [hunk 6]
```diff
@@ -4,6 +4,8 @@ title: What is Asylo?
 
 overview: Learn why Asylo could be right for your application.
 
+keywords: about, about asylo, what is asylo?, asylo, info, about enclave application.
```
Comment (KeithMoyer) [inline _about/about.md]: To keep this review discussion manageable, please refrain from customizing individual page keywords in this PR. Just adding the ability for per-page customization and updating the default list.

**Review by KeithMoyer — CHANGES_REQUESTED** [review]
Great! Things seem to be working now. :-)

File: _about/about.md (hunk 6 above)
Comment (aakashrajput) [reply _about/about.md]: so do I have to remove the keywords from other pages? or I just comment them?

Comment (shivaylamba) [reply _about/about.md]: Yes that is what we need to do. This PR is mainly geared towards adding just the functionality of page specific keyword and update default keywords.

Comment (aakashrajput) [reply _about/about.md]: reverted some changes, please check

File: _about/index.md [hunk 7]
```diff
@@ -2,6 +2,7 @@
 title: About Asylo
 overview: Get a bit more in-depth info about the Asylo project.
 
+
```
Comment (KeithMoyer) [inline _about/index.md]: The only change you're making to several files now is adding a blank line.

File: _about/name.md [hunk 8]
```diff
@@ -2,6 +2,7 @@
 title: Why the name Asylo?
 overview: Find out why the project name was chosen.
 
+
```
Comment (KeithMoyer) [inline _about/name.md]: here

File: _about/overview.md [hunk 9]
```diff
@@ -4,6 +4,7 @@ title: Asylo Overview
 
 overview: Get a better understanding of the problem space and how Asylo is designed to enable solutions.
 
+
```
Comment (KeithMoyer) [inline _about/overview.md]: here

File: _blog/2018/announcing-asylo.md [hunk 10]
```diff
@@ -34,4 +34,3 @@ get started with our
 [quickstart guide]({{home}}/docs/guides/quickstart.html) and dive into
 the [documentation]({{home}}/docs/). Be sure to check our website at
 [https://asylo.dev/]({{home}}/) for regular updates.
-
```
Comment (KeithMoyer) [inline _blog/2018/announcing-asylo.md]: only change you're making to this file is removing a line

File: _docs/reference/index.md [hunk 11]
```diff
@@ -1,6 +1,7 @@
 ---
 title: Reference
 overview: The Reference section contains detailed authoritative reference material such as C++ API, and message-oriented API calling parameters.
+
```
Comment (KeithMoyer) [inline _docs/reference/index.md]: here

File: _layouts/base.html [hunk 12]
```diff
@@ -3,6 +3,7 @@
 ---
 {% include home.html %}
 
+
```
Comment (KeithMoyer) [inline _layouts/base.html]: here

**Review by KeithMoyer — CHANGES_REQUESTED** [review]
Looking pretty good now. Just need to get rid of the unnecessary whitespace changes.

**Review by aakashrajput — COMMENTED** [review]
done

**Review by KeithMoyer — APPROVED** [review]
Thanks!
